### PR TITLE
CDAP-4231 - Refactor ETLStage to include name

### PR DESCRIPTION
**Pull request by gsps1**
Refactoring ETLStage to include a label name that is unique in a pipeline and extracting the plugin related information {type, properties and artifactId} into a separate class.
unique name and plugin object is used to represent an ETLStage.

Since this PR breaks the compatibility for UI, we will review and close this PR, while we have a working UI based on this branch, we can merge them together into develop.

This PR mainly contains the API change, unique name check while registering the pipeline, and related test changes.

JIRA: https://issues.cask.co/browse/CDAP-4231

### CODE REVIEW

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/dataset/DataSetInputFormat.java [hunk 1]
```diff
@@ -16,13 +16,13 @@
 
 package co.cask.cdap.internal.app.runtime.batch.dataset;
 
+import co.cask.cdap.api.common.Bytes;
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/dataset/DataSetInputFormat.java]: unintended?

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/TransformContext.java [hunk 2]
```diff
@@ -44,9 +44,9 @@
   StageMetrics getMetrics();
 
   /**
-   * Gets the stage number of the transform, useful for setting the context of logging in transforms.
+   * Gets the unique stage name of the transform, useful for setting the context of logging in transforms.
    *
    * @return stage number
    */
-  int getStageId();
+  String getStageId();
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/TransformContext.java]: lets use consistent terminology and call it getStageName()

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/TransformContext.java [hunk 3]
```diff
@@ -44,9 +44,9 @@
   StageMetrics getMetrics();
 
   /**
-   * Gets the stage number of the transform, useful for setting the context of logging in transforms.
+   * Gets the unique stage name of the transform, useful for setting the context of logging in transforms.
    *
    * @return stage number
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/TransformContext.java]: update

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/etl/batch/BatchCubeSinkTestRun.java [hunk 4]
```diff
@@ -72,11 +74,12 @@ public void test() throws Exception {
     Map<String, String> measurementsProps = ImmutableMap.of(
       Properties.Cube.MEASUREMENT_PREFIX + "count", "COUNTER"
     );
-    ETLStage sink = new ETLStage("Cube",
+    Plugin sinkConfig = new Plugin("Cube",
                                  ImmutableMap.of(Properties.Cube.DATASET_NAME, "batch_cube",
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/etl/batch/BatchCubeSinkTestRun.java]: nit: indentation

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/etl/batch/ETLEmailActionTestRun.java [hunk 5]
```diff
@@ -55,15 +56,22 @@ public void beforeTest() {
...
+    Plugin transformConfig = new Plugin("Projection", ImmutableMap.<String, String>of());
+
+    ETLStage source = new ETLStage("source", sourceConfig);
+    ETLStage sink = new ETLStage("sink", sinkConfig);
+    ETLStage transform = new ETLStage("transform", transformConfig);
     List<ETLStage> transformList = Lists.newArrayList(transform);
-    ETLStage action = new ETLStage("Email", ImmutableMap.of(EmailAction.RECIPIENT_EMAIL_ADDRESS, "to@test.com",
+
+    Plugin actionConfig = new Plugin("Email", ImmutableMap.of(EmailAction.RECIPIENT_EMAIL_ADDRESS, "to@test.com",
                                                             EmailAction.FROM_ADDRESS, "from@test.com",
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/test/java/co/cask/cdap/etl/batch/ETLEmailActionTestRun.java]: nit: indentation

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/AbstractTransformContext.java [hunk 6]
```diff
@@ -42,7 +42,7 @@ public AbstractTransformContext(PluginContext pluginContext, Metrics metrics, Lo
     this.pluginContext = pluginContext;
     this.stageId = stageId;
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/AbstractTransformContext.java]: as mentioned above, lets use 'name' instead of 'id' to be consistent with our terminology.

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/DefaultStageMetrics.java [hunk 7]
```diff
@@ -28,9 +28,9 @@
   private final Metrics metrics;
   private final String prefix;
 
-  public DefaultStageMetrics(Metrics metrics, PluginID id) {
+  public DefaultStageMetrics(Metrics metrics, String stageId) {
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/DefaultStageMetrics.java]: similar comment about name

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/PipelineRegisterer.java [hunk 8]
```diff
@@ -159,10 +162,28 @@ public Pipeline registerPlugins(ETLConfig config, Class errorDatasetType, Datase
     return new Pipeline(sourcePluginId, sinksInfo, transformInfos);
   }
 
+  private void validateStageNames(ETLStage sourceConfig, List<ETLStage> transforms, List<ETLStage> sinks) {
+    Set<String> uniquieStageNames = new HashSet<>();
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/PipelineRegisterer.java]: nit: typo in 'uniquie'

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/PipelineRegisterer.java [hunk 9]
```diff
@@ -159,10 +162,28 @@ public Pipeline registerPlugins(ETLConfig config, Class errorDatasetType, Datase
...
+    uniquieStageNames.add(sourceConfig.getName());
+    validateUniqueStageName(uniquieStageNames, transforms);
+    validateUniqueStageName(uniquieStageNames, sinks);
+  }
+
+  private void validateUniqueStageName(Set<String> uniqueStageNames, List<ETLStage> stages) {
+    for (ETLStage stage : stages) {
+      if (uniqueStageNames.contains(stage.getName())) {
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/PipelineRegisterer.java]: Set.add() returns a boolean, so you could use that instead of checking then adding.

**Comment by albertshau** [comment]
mostly small style comments.  Please add a test case for duplicate stage names, then lgtm.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/dataset/DataSetInputFormat.java (hunk 1 above)
Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/dataset/DataSetInputFormat.java]: not sure why it was showing as diff.. checked out develop for this file..

**Comment by gsps1** [comment]
@albertshau have addressed the comments. please look at it again! Thanks

**Comment by albertshau** [comment]
:+1:

**Comment by gsps1** [comment]
Have added backward compatibility, so this can be merged to develop and the code added for backward compatibility can be removed once UI is updated.
